### ISSUE TITLE
Add docs for using secrets with Dagger

### ISSUE THREAD

**Issue by d3rp3tt3** [dagger/dagger#4435]
## Summary
There are [several items in discovery now for how to improve how Dagger works with secrets](https://github.com/dagger/dagger/issues/4366). This issue clarifies why we need documentation for working with secrets and provides a rough scaffold for how that content.

### Why this matters
> Secrets support in Dagger allows you to utilize confidential information -- such as passwords, API keys, SSH keys, etc -- when running your Dagger Plans, *without*
 exposing those secrets in plaintext logs, writing them into the filesystem of containers you're building, or inserting them into cache.
> 

Source: Previous Dagger Secrets documentation [https://docs.dagger.io/1204/secrets/](https://docs.dagger.io/1204/secrets/)

## Problem
A previous version of the documentation included a section for secrets: [https://docs.dagger.io/1204/secrets/](https://docs.dagger.io/1204/secrets/). These docs, however, needed to be updated or carried over when the new version of the Dagger engine and the SDKs were launched.

Without prescriptive guidance, each person using Dagger today must figure out how to do this themselves. As a result, they may be putting confidential information directly into their pipelines.

## Solution
We should add docs for the *current* Dagger documentation to assist implementors using secrets in Dagger pipelines.

The core concepts to cover are:

- What secrets are and why they’re essential to keep private
- How to use secrets with Dagger pipelines, including examples for each SDK
- Best practices
- Roadmap/what’s next for additional features (I think this is optional, but I encourage the team sharing. This could also go into the top-level section for secrets and include a link to the current GitHub Issue - [[#4366](https://github.com/dagger/dagger/issues/4366)](https://github.com/dagger/dagger/issues/4366))

### Potential outline
This outline assumes the content would be put into the existing information architecture (IA) for the Dagger docs. 

[Dagger Docs]

- Introduction
- Secrets (What secrets are and why they must be kept private)
    - Best practices (cycling secrets, never use credentials in Dagger pipelines in plaintext, etc.)
    - How to use secrets
        - Environment variables
            - From the host machine
            - Into a container
        - Files

(I do support updating that IA to reduce repetition, make it easy to try Dagger quickly, and to use a just-in-time docs model so that developers don’t have to dig around in the docs to find what they need. However, that’s a separate topic being discussed in other GitHub issues, Discord, and the GitHub Discussions forum.)

## Acceptance criteria
- [ ]  Review existing secrets documentation and determine what can be brought over to the current docs
- [ ]  Add a new top-level section to the current Dagger documentation for secrets
- [ ]  In this Secrets section, cover best practices and how to use secrets
- [ ]  Provide an example for each method for each SDK (that is: using environment variables should be documented for every SDK and not just Go)

## Other supporting resources
- Previous secrets documentation: [https://docs.dagger.io/1204/secrets/](https://docs.dagger.io/1204/secrets/)
- https://github.com/dagger/dagger/issues/4366

**Comment by d3rp3tt3** [comment]
I will also look for any community-contributed examples of using secrets in Dagger pipelines. I think that there's one in a community call.

**Comment by shykes** [comment]
Thanks Jenn! wdyt @vikram-dagger ?

**Comment by vikram-dagger** [comment]
> Thanks Jenn! wdyt @vikram-dagger ?

I agree that we need to add more information to our docs on dealing with secrets. In terms of the IA, I would go a slightly different route though. Consider that our current documentation structure is segmented by components (like SDK, API, CLI...) and not by feature. So I worry that a top-level section for "Secrets" might not fit very well into the current structure. 

Barring an update to the structure (which we may well need to implement in future anyway as the content grows), I would propose instead:
- One or more guides with overview, explanation and code for how to use secrets (variables, files, external vaults, etc). This could be a separate guide for each SDK, or a single guide with multi-language examples. These would be hosted in the "Guides" sections that we currently have.
- A new top-level "Best Practices" section, which would cover best practices for using Dagger (including, but not necessarily limited to, secrets).  
- Additional entrypoints to this topic as Q&A related to secrets in the FAQ, with links to the above guides and best practice sections as needed.

**Comment by d3rp3tt3** [comment]
@vikram-dagger , I think that makes sense. You could also add partials for the reusable content.

**Comment by mircubed** [comment]
Closing #4501 since it is related to this, but this issue has way more context in it.

**Comment by mircubed** [comment]
@vikram-dagger , @dolanor got #4667 merged, so I think we can start to move the priority on this one. Let's do a guide and then a blog post
